dateutil/parser.py

```diff
@@ -1361,6 +1270,132 @@ def parse(self, tzstr):
 def _parsetz(tzstr):
     return DEFAULTTZPARSER.parse(tzstr)
 
+class InvalidDatetimeError(ValueError): pass
+class InvalidDateError(InvalidDatetimeError): pass
+class InvalidTimeError(InvalidDatetimeError): pass
+
+class ProgrammingError(AssertionError):
+    """ProgrammingError is for if we reach a state that we *think* should
+    be unreachable and would otherwise be inclined to assert is impossible.
+    Since an AssertionError is useful a user, a ProgrammingError clarifies
+    that they should file a bug report.
+    """
+
+
+def _parse_hms(i, l, info, res):
+    # TODO: This is still a mess.  Can we make this recursive instead of
+    # using a while loop?
+    len_l = len(l)
+
+    if ((i+1 < len_l and info.hms(l[i+1]) is not None)
+        or (i+2 < len_l and l[i+1] == ' ' and info.hms(l[i+2]) is not None)
+            ):
+        # HH[ ]h or MM[ ]m or SS[.ss][ ]s
+
+        value_repr = l[i]
+        value = float(value_repr)
+
+        if l[i+1] == ' ':
+            i += 1
+
+        idx = info.hms(l[i+1])
+        while True:
+            if idx == 0:
+                res.hour = int(value)
+                if value % 1:
+                    res.minute = int(60*(value % 1))
+
+            elif idx == 1:
+                (res.minute, res.second) = _parse_min_sec(value)
+
+            elif idx == 2:
+                (res.second, res.microsecond) = _parsems(value_repr)
+
+
+            if i+2 >= len_l or idx == 2:
+                i += 1
+                break
+
+            # 12h00
+            try:
+                value_repr = l[i+2]
+                value = float(value_repr)
+            except ValueError:
+                i += 1
+                break
+            else:
+                idx += 1
+
+                if i+3 < len_l:
+                    newidx = info.hms(l[i+3])
+
+                    if newidx is not None:
+                        idx = newidx
+                i += 2
+    return i
+
+
+# TODO: requre len(token) >= 3 like we do for the between-parens version?
+# do some other validation here instead of putting it off?  As of now, "Q"
+# will be accepted as a timezone...
+def _could_be_tzname(hour, tzname, tzoffset, token):
```

Comment:
require is misspelled here.

Also, we absolutely cannot require `len(token) >= 3`, since `Z` is in the ISO8601 spec for Zulu time (UTC), unless we want to special-case this.

That said, when we fix this up, we may well be able to determine exactly the minimum and maximum length of all valid time zones. We can start by checking the length of all the keys of `tzinfos`.